lib/ood_core/job/adapters/ccq.rb

```diff
@@ -0,0 +1,300 @@
+        # @param id [#to_s] the id of the job
+        # @raise [JobAdapterError] always
+        # @return [void]
+        def release(_)
+          raise NotImplementedError, "subclass did not define #release"
+        end
+
+        # Delete the submitted job
+        # @param id [#to_s] the id of the job
+        # @return [void]
+        def delete(id)
+          call("ccqdel", args: [id])
+        end
+
+        def directive_prefix
+          '#CC'
+        end
+
+        private
+
+        # Mapping of state codes
+        STATE_MAP =
+        {
+          'Error'         => :suspended,  # not running, but infrastructure still possibly exists
+          'CreatingCG'    => :queued,     # creating control group
+          'Pending'       => :queued,     # in queue
+          'Submitted'     => :queued,     #
+          'Provisioning'  => :queued,     # node is being provisioned
+          'Running'       => :running,    #
+          'Completed'     => :completed,  #
+        }.freeze
+
+        def gcp_provider
+          'gcp'
+        end
+
+        def aws_provider
+          'aws'
+        end
+
+        def image_arg
+          if cloud == gcp_provider
+            '-gcpgi'
+          else
+            '-awsami'
+          end
+        end
+
+        def call(cmd, args: [], env: {}, stdin: "")
+          cmd = OodCore::Job::Adapters::Helper.bin_path(cmd, bin, bin_overrides)
+          args = args.map(&:to_s)
+          env = env.to_h
+          o, e, s = Open3.capture3(env, cmd, *args, stdin_data: stdin.to_s)
+          s.success? ? o : interpret_and_raise(e, cmd)
+        end
+
+        # helper function to interpret an error the command had given and
+        # raise a different error.
+        def interpret_and_raise(error, command)
+          # a special case with CCQ that prompts the user for username & password
+          # so let's be helpful and tell the user what to do.
+          if error.end_with?("EOFError: EOF when reading a line\n")
+            raise(
+              PromptError,
+              "The #{command} command was prompted. You need to generate the certificate " + 
+              "manually in a shell by running 'ccqstat'\nand entering your username/password"
+            )
+          else
+            raise(JobAdapterError, e.message)
+          end
+        end
+
+        # Convert seconds to duration
+        def seconds_to_duration(seconds)
+          format("%02d:%02d:%02d", seconds / 3600, seconds / 60 % 60, seconds % 60)
+        end
+
+        # helper to make a script file. We can't pipe it into ccq so we have to
+        # write a file.
+        def make_script_file(content)
+          file = Tempfile.new(tmp_file_name)
+          file.write(content.to_s)
+          file.flush
+          file
+        end
+
+        def tmp_file_name
+          'ccq_ood_script_'
+        end
+
+        def parse_job_id_from_ccqsub(output)
+          match_data = /job id is: (?<job_id>\d+) you/.match(output)
```

Comment:
This looks brittle in several respects.

1. will the job id always be 2 or more digits of an integer value (never starting with 1, 2, 3), regardless of whether Torque or Slurm is used?
2. will ccqsub always output the job id in between the words "job id is: " and " you"?

We had/have something similar to parse the job ID from LSF bsub output and it was problematic between different LSF versions. 

It might be worth requesting the CCQ developers to add to ccqsub support for a new argument that ensures only the job id is printed to stdout, instead of all of the human readable language around the job id

FWIW I did notice in the example text the only digits were either in the name of the host (slurm1) or the job id itself (123456). So perhaps the regex could be simpler if we knew more about it. Or maybe the regex used should actually be configurable - something that we can put in the cluster config, with a sane default - so at least our guess can be fixed without a code change if a situation arises.

Reply:
So we spoke with the CCQ team and they indicated they'd at least keep this stable and backward compatible.  We're slightly lucky in that we can release atomically with them when they and it's cloud so folks will just replace instead of upgrade.

Either way I've made it configurable, but we're still looking for the named group `job_id` so there's a little extra in there in case someone sets the regex to look for `jobid` or similar, it'll at least throw an error that indicates the issue and not an uncaught exception.